### ISSUE TITLE
End damping regressed to a dt-scaled rate in 79064c4d: three linear cases blow up, kbm gamma off 65% (shipped in 2.0.0)

### ISSUE THREAD

**Issue by rogeriojorge** [uwplasma/GKX#192]
`tools/comparison/fixtures/parity/w7x_itg.toml`, run at the settings `tools/gx_parity_matrix_manifest.toml` pins for `w7x_itg` (Nl=8, Nm=16, dt=0.005, `imex2`, 40000 steps, t_max=200), reaches `|field|` ≈ 5.75e+75. The matrix records finite gamma/omega for this case, so the recorded numbers cannot have come from a run at these settings — or something has changed since they were recorded.

## What is ruled out

**It is not a changed equilibrium.** The fixture takes its geometry from the comparison code's own output (`model = "eik"`, `geometry_file = "$GX_PARITY_REF_DIR/ITG_w7x/itg_w7x_adiabatic_electrons.out.nc"`), and `~/gx_refs/` on the office box is empty, so the reference had to be regenerated. A regenerated equilibrium would be the obvious explanation, since the equilibrium sets the stability bound.

It is not that. A freshly generated output and the `gx_rebaseline_precsqrt_20260819` copy are identical on the quantities that set the bound:

| | fresh | 20260819 rebaseline |
|---|---|---|
| max abs gradpar | 0.013013954274356365 | 0.013013954274356365 |
| bmag max/min | 1.390105 | 1.390105 |
| nz | 256 | 256 |

**It is not immediate.** Truncated to t_max=4.0 (800 steps at the pinned dt) on CPU in float64, the case completes cleanly: `gamma=0.189658 +/- 0.015587`, `omega=0.393486 +/- 0.013868`, all finite. The warnings it emits are the expected under-resolution ones for a short horizon. So this is a late-time instability, not a configuration that fails from the first step.

## What is not yet known

- Whether this is a regression or has always been true of these pinned settings. The decisive test is the same case at the same settings on an older commit; it was not run, because both office GPUs are saturated by unrelated jobs and GPU 0 now refuses allocation with `CUDA_ERROR_OUT_OF_MEMORY`, leaving only slow CPU runs.
- When in the trace it diverges.
- Whether `imex2` is the relevant factor. `dt` is fixed here, and `max|gradpar|` is 0.013, which is small — the parallel-streaming bound that explains the other decks in this repo does not obviously apply.

## Why this is not release-blocking for 2.0.0

The release gates that read this manifest are structural: they check that a `build_reproducibility_floor` is declared only where one was measured, that the builder reads the declared floor, and that the KBM floor covers its measured two-build spread. None of them runs a case or asserts a recorded gamma. All ten release checkers pass.

This fixture also lives under `tools/`, is a development comparison input, and is not a shipped example. `_CFL_MARGIN_OVER_BOUND` is currently empty — no shipped deck is recorded over its bound.

## Note on provenance

The manifest records `reference_provenance = "generated by this campaign; no upstream reference ships for this case"` for `w7x_itg`. A self-generated reference that cannot be reproduced from its recorded inputs is the same failure mode as #178 and #173: a number in the repository that no command regenerates.

**Comment by rogeriojorge** [comment]
## The case in this issue is not the case that blows up

`w7x_itg` at the pinned settings **completes on current `main`**. The `5.75e+75` in the report belongs to `cyclone_salpha_itg`.

Run on `origin/main` d4bc276f, `~/venvs/dkx-gpu` (jax 0.10.2), `JAX_ENABLE_X64=1`, references in `/home/rjorge/gx_refs_lin`, everything at the manifest's pinned values:

| case | pinned | result on d4bc276f |
|---|---|---|
| `w7x_itg` | Nl=8, Nm=16, dt=0.005, imex2, 40000 steps | **completes**, settled 22/27 ky, peak ky=1.6 dγ=+3.68e-5 dω=-5.54e-6 |
| `cyclone_salpha_itg` | Nl=16, Nm=48, dt=0.002, imex2, 75000 steps | `FloatingPointError`, \|field\| **5.75e+75** |
| `cyclone_miller_itg` | Nl=16, Nm=48, dt=0.002, imex2, 75000 steps | `FloatingPointError`, non-finite field history |
| `cyclone_miller_kinetic_electrons` | Nl=16, Nm=48, dt=2e-4, imex2, 100000 steps | `FloatingPointError`, \|field\| 6.56e+90 |
| `kbm_miller` | Nl=16, Nm=48, dt=2e-4, imex2, 100000 steps | completes, but γ is off by −85 % … +32 % |
| `hsx_itg` | — | cannot be run at all, see below |

`w7x_itg` reproduces the recorded matrix at every ky: the settled band agrees to 1e-5 or better, and the summary is `max|dγ|(settled)` 0.031708 against the recorded 0.031614. So the recorded numbers for this case *are* reproducible from the recorded inputs, and the provenance concern raised in the issue does not apply to `w7x_itg`.

The four cases that misbehave are exactly the four with `nperiod = 2` (ntheta=32, Nz=96). The two with `nperiod = 1` (ntheta=256, Nz=256) are `w7x_itg`, which is exact, and `hsx_itg`, which cannot be run.

## It is a regression, and it is one commit

| commit | | `cyclone_salpha_itg` at pinned settings |
|---|---|---|
| e89c7fed | Release GKX 1.8.2 | **PASS** — bit-identical to `docs/_static/gkx_gx_linear_parity_matrix.json` |
| ab838be2 | core: consolidate Hermitian ky projection | **PASS** — bit-identical to the recorded matrix |
| 5fd8b770 | merge of #149 | adds no `src/gkx` change over ab838be2 (`git diff --stat ab838be2 5fd8b770 -- src/gkx` is empty) |
| **79064c4d** | **fix: keep end damping independent of step size** | **FAIL**, \|field\| 5.75e+75 |
| 7811610d | linear: support implicit diagnostic sampling | FAIL, \|field\| 5.75e+75 |
| d4bc276f | main | FAIL, \|field\| 5.75e+75 |

"Bit-identical" is literal: at v1.8.2 and at ab838be2 every γ and ω at all 11 ky equals the recorded value to the last digit, and so does every recorded relative difference. The GX side contributes nothing — the reference regenerated today for this case is bit-identical to the one behind the recorded matrix, same 34-sample trace, same `t_end = 150.00393876064894`.

Since the merge adds no source change, 79064c4d is exactly the end-damping edit applied to a tree that passes. Single-commit attribution, no ambiguity.

## Mechanism

79064c4d deleted a `/dt` from `_scalar_params` in `src/gkx/terms/assembly.py`:

```python
 damp_amp = jnp.asarray(params.damp_ends_amp, dtype=real_dtype)
-if dt is not None:
-    dt_arr = jnp.asarray(dt, dtype=real_dtype)
-    damp_amp = jnp.where(dt_arr != 0.0, damp_amp / dt_arr, damp_amp)
```

`damp_amp` feeds `end_damping_contribution`, which is an RHS term, so it is multiplied by `dt` in the step. Before the change the deck's `damp_ends_amp` was a fixed fraction removed **per step**; after it, `damp_ends_amp * dt` per step. Every deck therefore lost a factor `1/dt` of end damping, and none of the parity fixtures was re-tuned — all six still carry `damp_ends_amp = 0.1` and `[terms] end_damping = 1.0`. The loss orders the damage exactly:

| dt | damping lost | cases | outcome |
|---|---|---|---|
| 0.005 | ×200 | `w7x_itg` | still correct |
| 0.002 | ×500 | `cyclone_salpha_itg`, `cyclone_miller_itg` | blow up |
| 0.0002 | ×5000 | `cyclone_miller_kinetic_electrons`, `kbm_miller` | blow up / γ off by −65 % |

### Restoring the factor on `main` recovers the recorded numbers exactly

Rerun of `cyclone_salpha_itg` on d4bc276f at the pinned settings, with `damp_ends_amp` rescaled in memory from 0.1 to `0.1/dt = 50.0` and nothing else changed (no file in the repository touched — the config dataclass is replaced in the process):

- it completes, no overflow;
- every γ and ω at all 11 ky matches the recorded matrix to **6.9e-17 in γ (one ULP) and exactly in ω**.

So the missing `1/dt` is not merely correlated with the failure — putting it back on current `main` reproduces the recorded spectrum bit-for-bit.

## Where in the trace it diverges

Rerun on `main` with the finite-history guard disabled in-process so the whole trace survives; `phi_t` is (1875 samples, 11 ky, 1 kx, 96 z).

**Only ky = 0.55, the top of the scan, diverges.** The other ten ky keep their correct physical growth rate all the way to t = 150 — local `d(ln|phi|)/dt` at t = 120 against the recorded γ:

| ky | 0.10 | 0.15 | 0.20 | 0.25 | 0.30 | 0.35 | 0.40 | 0.45 | 0.50 |
|---|---|---|---|---|---|---|---|---|---|
| local | .0324 | .0548 | .0752 | .0884 | .0931 | .0897 | .0807 | .0685 | .0537 |
| recorded | .0306 | .0549 | .0750 | .0884 | .0930 | .0899 | .0808 | .0686 | .0542 |

So the physics is intact; one runaway mode aborts the whole batched case through a guard that is evaluated on the batch as a whole.

Onset is sharp at **t ≈ 49.5** — ky=0.55's local growth rate jumps from +0.018 to +2.40 between t = 49.36 and t = 49.52. The spurious rate is +2 to +3, roughly 25× the physical 0.093. ky = 0.50 starts going the same way around t ≈ 135.

Amplitude crossings of the global peak: 1e-6 at t=54.0, 1e0 at t=61.1, 1e6 at t=68.0, 1e12 at t=75.2, 1e20 at t=84.6, **1e30 (the guard threshold) at t=96.1**, 1e75 at t=149.1, final 5.748846451623465e+75. The manifest's fit window is [105, 150]; by t = 105 the field is already ~1e38, so the window sits entirely inside the blown-up region.

**Where in z:** before onset ky=0.55 peaks at z index 46–50, mid-domain — the normal ballooning eigenfunction, the same shape as healthy ky=0.30. From t = 49.5 the argmax jumps to z = 85–91 and z = 7–11 and stays there: **both ends of the 96-point parallel domain**. By t = 54 the normalized z profile is 0.74 / 0.73 in the first two bins and 0.50 / 1.00 in the last two, with mid-domain suppressed to 5e-3. End damping is precisely the term that acts there.

## What is ruled out

**Not the CFL step.** Using GKX's own `_linear_frequency_bound` at each case's pinned Nl/Nm (imex2, `cfl_fac` 1.0, `cfl` 0.9):

| case | dt | CFL-stable dt | dt/bound | over CFL? | outcome |
|---|---|---|---|---|---|
| `cyclone_salpha_itg` | 0.002 | 0.0038429 | 0.520 | no | blows up |
| `cyclone_miller_itg` | 0.002 | 0.0033350 | 0.600 | no | blows up |
| `cyclone_miller_kinetic_electrons` | 0.0002 | 0.00024096 | 0.830 | no | blows up |
| `kbm_miller` | 0.0002 | 0.00024658 | 0.811 | no | γ off by −65 % |
| `w7x_itg` | 0.005 | 0.0058909 | 0.849 | no | correct |

No case exceeds its own bound, and the ordering is backwards: the case with the **largest** margin blows up, the one with the smallest is fine. The error message's advice ("reduce dt") is not the diagnosis — the step is already at half the estimated stable step. (Cross-check: this computation reproduces the issue's `max|gradpar| = 0.013013954274356365` for w7x exactly.)

**Not precision.** The float32 sweep blows up on the same cases (non-finite rather than 5.75e+75, i.e. sooner). Both float64 and float32 fail, so this is not a rounding effect.

**Not the device.** An independent float64 run on the CPU backend (`JAX_PLATFORMS=cpu`) diverges the same way — crossings at t = 54.24 / 61.36 / 68.40 / 75.52 / 84.88 / 96.72 against the GPU's 54.00 / 61.12 / 68.00 / 75.20 / 84.56 / 96.08. Same onset, same ky, same z structure.

**Not the equilibrium**, as the issue already established, and not the deck: the only change to `tools/comparison/fixtures/parity/cyclone_salpha_itg.toml` since the matrix was recorded is the deletion of `use_diffrax = false`, and the manifest's pinned settings are unchanged.

## Separately: `hsx_itg` is the case with no generator

`hsx_itg` cannot be run at all. There is no HSX reference output, no GX input deck, and no VMEC `wout` anywhere on the box — `find ~ -name "wout*hsx*" -o -name "itg_hsx*"` returns nothing, `gx_refs_lin` holds only `ITG_cyclone`, `ITG_w7x`, `KAW`, `KBM`, and upstream ships no HSX case. Its recorded row in the parity matrix is therefore not reproducible from anything that exists. That is the failure mode the issue's provenance note describes, and it lands on `hsx_itg` rather than on `w7x_itg`.

## Not addressed here

Whether the pre-79064c4d or the post-79064c4d convention is the physically correct one is a separate question from this regression. What is established is that all six parity fixtures were calibrated under the old convention, none was re-tuned, and four of them no longer integrate.

No timing or throughput comparison was attempted: both office GPUs were saturated by unrelated jobs throughout, so wall-clock here would be meaningless.

**Comment by rogeriojorge** [comment]
**Retitle needed — the blow-up is not W7-X.** I attributed the 5.75e+75 signature to `w7x_itg` when I opened this. That was wrong, and the correction matters because it points at a much larger defect.

`w7x_itg` at the pinned settings **completes on main and reproduces the recorded matrix** (settled 22/27 ky, peak ky=1.6 dgamma=+3.68e-5). The 5.75e+75 belongs to `cyclone_salpha_itg`.

## What is actually broken

At the manifest's pinned settings on main:

| case | dt | outcome |
|---|---|---|
| `cyclone_salpha_itg` | 0.002 | FAIL, \|field\| 5.75e+75 |
| `cyclone_miller_itg` | 0.002 | FAIL, non-finite field history |
| `cyclone_miller_kinetic_electrons` | 0.0002 | FAIL, \|field\| 6.56e+90 |
| `kbm_miller` | 0.0002 | completes, gamma off by −65% |
| `w7x_itg` | 0.005 | correct |
| `hsx_itg` | — | cannot be run at all: no reference, no deck, no wout exists anywhere |

## Cause: commit 79064c4d, single-commit attribution

`79064c4d` ("fix: keep end damping independent of step size") removed `damp_amp = damp_amp / dt` from the RHS scalar assembly. Since the step applies `G += dt * RHS`, the old code removed a fixed **fraction per step**; the new one applies a fixed **rate**, so damping is weaker by a factor of `dt` — 200× at dt=0.005, 500× at 0.002, 5000× at 0.0002. **Severity orders exactly by 1/dt.**

Bisect: `ab838be2` passes bit-identically to the recorded artifact, `79064c4d` fails, and `git diff ab838be2 5fd8b770 -- src/gkx` is empty.

**Causal confirmation:** rescaling `damp_ends_amp` in memory from 0.1 to `0.1/dt` = 50.0 on current main makes the case complete and reproduce the recorded artifact to **6.939e-17 in gamma (one ULP) and exactly in omega at all 11 ky**.

## Ruled out

- **Not CFL.** No case exceeds its bound. The case with the largest margin (salpha, 0.52×) fails; the smallest (w7x, 0.849×) passes. The bound does not order the failures.
- **Not precision.** float32 and float64 both fail, with the same wrong gamma on kbm.
- **Not a device artifact.** CPU and GPU diverge identically.
- **Not the reference.** The regenerated GX output for salpha is bit-identical to the one behind the recorded matrix; GKX at v1.8.2 is bit-identical to the recorded GKX values. The entire change is GKX-side.

## Impact

This shipped in **2.0.0**. 37 shipped decks enable end damping, and nonlinear decks are time-stepped by default. A fix is in progress for 2.0.1. Separately, `hsx_itg`'s recorded row is not reproducible from anything that exists on disk, which is the same class of problem as #178.
